lib/Basic/Statistic.cpp

```diff
@@ -370,26 +518,72 @@ UnifiedStatsReporter::saveAnyFrontendStatsEvents(
+  // or profilers, we're not tracing or profiling: return early.
   if (!LastTracedFrontendCounters)
     return;
   auto Now = llvm::TimeRecord::getCurrentTime();
+
+  if (EventProfilers) {
+    auto TimeDelta = Now;
+    TimeDelta -= EventProfilers->LastUpdated;
+    EventProfilers->UserTime.recordEvent(T.EventName,
+                                         TimeDelta.getUserTime(),
+                                         IsEntry);
+    EventProfilers->SystemTime.recordEvent(T.EventName,
+                                           TimeDelta.getSystemTime(),
+                                           IsEntry);
+    EventProfilers->ProcessTime.recordEvent(T.EventName,
+                                            TimeDelta.getProcessTime(),
+                                            IsEntry);
+    EventProfilers->WallTime.recordEvent(T.EventName,
+                                         TimeDelta.getWallTime(),
+                                         IsEntry);
+    EventProfilers->LastUpdated = Now;
+  }
+
+  if (EntityProfilers) {
+    auto TimeDelta = Now;
+    TimeDelta -= EntityProfilers->LastUpdated;
+    EntityProfilers->UserTime.recordEvent(T.EventName,
+                                          TimeDelta.getUserTime(),
+                                          IsEntry, T.Entity, T.Formatter);
+    EntityProfilers->SystemTime.recordEvent(T.EventName,
+                                            TimeDelta.getSystemTime(),
+                                            IsEntry, T.Entity, T.Formatter);
+    EntityProfilers->ProcessTime.recordEvent(T.EventName,
+                                             TimeDelta.getProcessTime(),
+                                             IsEntry, T.Entity, T.Formatter);
+    EntityProfilers->WallTime.recordEvent(T.EventName,
+                                          TimeDelta.getWallTime(),
+                                          IsEntry, T.Entity, T.Formatter);
+    EntityProfilers->LastUpdated = Now;
+  }
+
   auto StartUS = uint64_t(1000000.0 * T.SavedTime.getProcessTime());
   auto NowUS = uint64_t(1000000.0 * Now.getProcessTime());
   auto LiveUS = IsEntry ? 0 : NowUS - StartUS;
   auto &C = getFrontendCounters();
-#define FRONTEND_STATISTIC(TY, NAME)                          \
-  do {                                                        \
-    int64_t total = C.NAME;                                    \
-    int64_t delta = C.NAME - LastTracedFrontendCounters->NAME; \
-    static char const *name = #TY "." #NAME;                  \
-    if (delta != 0) {                                         \
-      LastTracedFrontendCounters->NAME = C.NAME;              \
-      FrontendStatsEvents.emplace_back(FrontendStatsEvent {   \
-          NowUS, LiveUS, IsEntry, T.EventName, name,          \
-            delta, total, T.Entity, T.Formatter});            \
-    }                                                         \
+#define FRONTEND_STATISTIC(TY, NAME)                            \
```

Comment:
This was actually the macro that scared me most, but I guess it's hard to factor out.

```c++
#define FRONTEND_STATISTIC(TY, NAME) \
  handleStatistic(#TY "." #NAME, C.NAME, LastTracedFrontendCounters->NAME, EventProfilers->NAME, EntityProfilers->NAME);
```

…well, maybe?

Reply:
Mm, fair. Lemme see what I can do.